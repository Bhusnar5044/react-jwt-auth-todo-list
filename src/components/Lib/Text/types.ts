import { FontVariants, FontWeights } from '@theme/font/types';
import { HTMLProps } from '@utils/types';

interface CommonProps extends HTMLProps<HTMLElement> {
    textColor?: string;
    uppercase?: boolean;
    lineThrough?: boolean;
    fullWidth?: boolean;
    as?: keyof JSX.IntrinsicElements | React.ComponentType<any>;
}

export interface TextProps extends CommonProps {
    textVariant?: FontVariants;
    textWeight?: FontWeights;
    textAlign?: 'initial' | 'left' | 'right' | 'center';
}
