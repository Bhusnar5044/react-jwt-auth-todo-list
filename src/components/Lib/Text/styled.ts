import { theme } from '@theme';
import { WithThemeProp } from '@utils/types';
import styled, { css } from 'styled-components';
import { TextProps } from './types';

const uppercase = () => css`
    text-transform: uppercase;
`;

const lineThrough = () => css`
    text-decoration: line-through;
`;
const getTextStyle = (props: TextProps) => {
    const { textColor, textVariant, textWeight, textAlign } = props,
        { defaults, weights, variants } = theme.font,
        { fontSize, fontWeight, lineHeight, letterSpacing } = variants[textVariant || defaults.variant];

    return css`
        margin: 0;
        color: ${textColor};
        font-size: ${fontSize};
        font-weight: ${weights[textWeight || fontWeight]};
        letter-spacing: ${letterSpacing};
        line-height: ${lineHeight};
        text-align: ${textAlign};
    `;
};

export const TextStyled = styled('span')<TextProps & WithThemeProp>`
    ${getTextStyle}

    ${({ fullWidth, theme }) =>
        fullWidth &&
        css`
            display: block;
            margin: ${theme.spacing.S2} 0;
        `}

    ${(props) => props.uppercase && uppercase()};
    ${(props) => props.lineThrough && lineThrough()};
    display: ${({ as }) => (as === 'strong' || as === 'span' || as === 'a') && 'inline-block'};
`;

TextStyled.defaultProps = {
    theme: theme,
};
