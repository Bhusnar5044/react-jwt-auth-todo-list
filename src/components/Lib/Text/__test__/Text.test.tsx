import { render, screen } from '@test-utils';
import { Text } from '../Text';

describe('Text', () => {
    it('should render properly', () => {
        const { container } = render(<Text />);
        expect(container).toMatchSnapshot();
    });

    it('should render properly with content provided', () => {
        render(<Text>Dummy Text</Text>);
        expect(screen.getByText(/Dummy Text/i)).toBeInTheDocument();
    });

    it('should render properly with content provided', () => {
        render(<Text>Dummy Text</Text>);
        expect(screen.getByText(/Dummy Text/i)).toBeInTheDocument();
    });

    it('should render properly with provided style props', () => {
        render(
            <Text textColor="red" textAlign="center">
                Dummy Text
            </Text>,
        );
        expect(screen.getByText('Dummy Text')).toHaveStyleRule('color', 'red', 'text-align', 'center');
    });

    it('should render properly with uppercase and lineThrough prop', () => {
        render(
            <Text uppercase lineThrough>
                Dummy Text
            </Text>,
        );
        const text = screen.getByText(/DUMMY TEXT/i);
        expect(text).toBeInTheDocument();
        expect(text).toHaveStyleRule('text-decoration', 'line-through');
    });
});
