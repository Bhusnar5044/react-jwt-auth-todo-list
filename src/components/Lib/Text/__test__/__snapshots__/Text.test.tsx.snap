// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text should render properly 1`] = `
.c0 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
  display: inline-block;
}

<div>
  <span
    class="c0"
    textalign="initial"
    textcolor="inherit"
  />
</div>
`;
